processor/k8sattributesprocessor/internal/kube/client.go

```diff
@@ -639,3 +657,71 @@ func (c *WatchClient) extractNamespaceLabelsAnnotations() bool {
 func needContainerAttributes(rules ExtractionRules) bool {
 	return rules.ContainerImageName || rules.ContainerImageTag || rules.ContainerID
 }
+
+func (c *WatchClient) handleReplicaSetAdd(obj interface{}) {
+	observability.RecordReplicaSetAdded()
+	if replicaset, ok := obj.(*apps_v1.ReplicaSet); ok {
+		c.addOrUpdateReplicaSet(replicaset)
+	} else {
+		c.logger.Error("object received was not of type apps_v1.ReplicaSet", zap.Any("received", obj))
+	}
+}
+
+func (c *WatchClient) handleReplicaSetUpdate(old, new interface{}) {
+	observability.RecordReplicaSetUpdated()
+	if replicaset, ok := new.(*apps_v1.ReplicaSet); ok {
+		c.addOrUpdateReplicaSet(replicaset)
+	} else {
+		c.logger.Error("object received was not of type apps_v1.ReplicaSet", zap.Any("received", new))
+	}
+}
+
+func (c *WatchClient) handleReplicaSetDelete(obj interface{}) {
+	observability.RecordReplicaSetDeleted()
+	if replicaset, ok := obj.(*apps_v1.ReplicaSet); ok {
+		c.m.Lock()
+		key := string(replicaset.UID)
+		delete(c.ReplicaSets, key)
+		c.m.Unlock()
+	} else {
+		c.logger.Error("object received was not of type apps_v1.ReplicaSet", zap.Any("received", obj))
+	}
+}
+
+func (c *WatchClient) addOrUpdateReplicaSet(replicaset *apps_v1.ReplicaSet) {
+	var deployment *Deployment
+	for _, ownerReference := range replicaset.OwnerReferences {
+		if ownerReference.Kind == "Deployment" {
+			deployment = &Deployment{
```

Comment:
This is hopefully a rare case, but can we add a warn log here if we find that another deployment owner reference has been found? I think that may cause strange and confusing data if it were to happen.

Alternatively, if we know for sure that Kubernetes guarantees a ReplicaSet will only ever have a single Deployment as an owner, then we should break out of the loop and have a comment documenting that there can only be one reference.

Reply:
After some research, I think we can use `ownerReference.controller` which denotes whether this controller take over the control of replicaset. The design doc is right [here](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/controller-ref.md)

I also made a test, I can create a replicaset which ownerReferences have two deployments in it, but only one `ownerReference.controller` is true.
```
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  generation: 1
  labels:
    app: test2
  name: test2-7c9d5954f6
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: Deployment
    name: test2
    uid: 2ab30db9-c151-4db4-bc43-680bfcfba696
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: false
    kind: Deployment
    name: test1
    uid: 13ca1c8b-5c96-4ec0-a3af-0e08110e22dc
```

If I set both controller to true, the apiserver will deny the request with error message.
```
Only one reference can have Controller set to true. Found "true" in references for test2 and test1
```